numpy/lib/_type_check_impl.py

```diff
@@ -680,20 +670,9 @@ def common_type(*arrays):
     <class 'numpy.complex128'>
 
     """
-    is_complex = False
-    precision = 0
-    for a in arrays:
-        t = a.dtype.type
-        if iscomplexobj(a):
-            is_complex = True
-        if issubclass(t, _nx.integer):
-            p = 2  # array_precision[_nx.double]
-        else:
-            p = array_precision.get(t, None)
-            if p is None:
-                raise TypeError("can't get common type for non-numeric array")
-        precision = max(precision, p)
-    if is_complex:
-        return array_type[1][precision]
-    else:
-        return array_type[0][precision]
+    common_type = _multiarray_umath.result_type(*arrays).type
```

Comment:
OK, this will make it a bit slower possibly, but I think I would like to:
1. Use `*(a.dtype for a in arrays)` because unfortunately `result_type` supports both array and dtype inputs which would broad the signature here.
2. We can add `0.0` as an additional argumen tto `result_type()` to enforce the `integer -> int64` transition.

Unfortunately, this will improve the situation when mixing integers and floats:  `float32 + int16` will actually promote to stay `float32`.  While here any integer forces float64 precision.

Reply:
1. This is to make sure people do not pass a `dtype` directly to `common_type`? It does make the code a bit slower, but the overhead result is still ok. 
2.  We could use `common_type = _multiarray_umath.result_type(0., *arrays).type` and skip the check on integer type. The pure int goes to float64 (as in current main), but a disadvantage is that now `float32+float32` goes to `float64` (where before it would be `float32`)

With 
```
def common_type_with_zero(*arrays):
    common_type = _multiarray_umath.result_type(0., *[a.dtype for a in arrays]).type
    if not issubclass(common_type, _nx.inexact):
        raise TypeError("can't get common type for non-numeric array")
    return common_type

i16=np.array(2, dtype=np.int16)
i32=np.array(2, dtype=np.int32)
f32=np.array(2, dtype=np.float32)

for c in [common_type_original,common_type_pr24467, common_type_pr, common_type_with_zero]:
    print(f'\n# {c}')
    print(f'c(i16, i32): {c(i16, i32)}' )
    print(f'c(i16, f32): {c(i16, f32)}' )
    print(f'c(f32, f32): {c(f32, f32)}' )
```
I get
```
# <function common_type_original at 0x000001EE32CE5F80>
c(i16, i32): <class 'numpy.float64'>
c(i16, f32): <class 'numpy.float64'>
c(f32, f32): <class 'numpy.float32'>

# <function common_type_pr24467 at 0x000001EE32CE5800>
c(i16, i32): <class 'numpy.float64'>
c(i16, f32): <class 'numpy.float64'>
c(f32, f32): <class 'numpy.float32'>

# <function common_type_pr at 0x000001EE32CE51C0>
c(i16, i32): <class 'numpy.float64'>
c(i16, f32): <class 'numpy.float32'>
c(f32, f32): <class 'numpy.float32'>

# <function common_type_with_zero at 0x000001EE32CE5DA0>
c(i16, i32): <class 'numpy.float64'>
c(i16, f32): <class 'numpy.float64'>
c(f32, f32): <class 'numpy.float64'>
```

Performance for main, PR24467, this PR and the variation with zero:
```
1.24 µs ± 79.8 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
740 ns ± 55.8 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
333 ns ± 16.4 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
998 ns ± 20 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
```
At this point I have a slight preference for PR 24467, but it is a bit comparing apples and oranges.

Reply:
To me, those results look like you ran it with `result_type(0, *arrays)` and not `result_type(0., *[a.dtype for a in arrays])`.  But yes, you still get the change for integers, which is maybe for the better, but is a change.

Reply:
This really was with `result_type(0., *[a.dtype for a in arrays])`. For `result_type(0, *[a.dtype for a in arrays])` the case `common_type(np.array([0]), np.array([0]))` results in a `TypeError`.

Reply:
@eendebakpt not sure it matters due to the other problems, but you wrote `0` rather than `0.` there, which is very different.  I just tried just to cover my ground triply: `common_type_with_zero` will give f32 for `f32, f32`.  (The problem is `i16, f32`)

Reply:
@seberg My mistake, I was working with the old promotion rules. The version you suggested works fine for `f32, f32`.

Updated output with benchmarks:
```
module 'numpy' from '/mnt/data/numpy/numpy/__init__.py'> 2.0.0.dev0+git20230831.20d03bc
weak

# <function common_type_original at 0x7fbf99303d90>
c(i16, i32): <class 'numpy.float64'>
c(i16, f32): <class 'numpy.float64'>
c(f32, f32): <class 'numpy.float32'>
c(i16, i32) 1.445 [dt]
c(f64, f64) 1.548 [dt]

# <function common_type_pr24467 at 0x7fbf8487f9a0>
c(i16, i32): <class 'numpy.float64'>
c(i16, f32): <class 'numpy.float64'>
c(f32, f32): <class 'numpy.float32'>
c(i16, i32) 1.055 [dt]
c(f64, f64) 0.886 [dt]

# <function common_type_pr at 0x7fbf8487fa30>
c(i16, i32): <class 'numpy.float64'>
c(i16, f32): <class 'numpy.float32'>
c(f32, f32): <class 'numpy.float32'>
c(i16, i32) 0.371 [dt]
c(f64, f64) 0.393 [dt]

# <function common_type_with_zero at 0x7fbf8487fac0>
c(i16, i32): <class 'numpy.float64'>
c(i16, f32): <class 'numpy.float64'>
c(f32, f32): <class 'numpy.float32'>
c(i16, i32) 0.960 [dt]
c(f64, f64) 0.951 [dt]
```
For `common_type_with_zero` I created #24668. It is faster than main and uses the `result_type` internally. Without the `*[a.dtype for a in arrays]` it would be about 30% faster (python 3.10). The only downside I found so far is the incompatibility with datetime64.

Reply:
@eendebakpt I am still confused at the list `np.result_type(0.0, np.int16, np.float32)` gives float32 which is a change in behavior, no (or maybe I am mixing up which version is which)?

I am almost tempted to live with that change of behavior since we have a lot of other promotion related changes, but I have to think about it a bit.

(I suppose my ask is just to confirm that it _is_ a change in behavior.  In a sense, the old behavior is more like `result_type(*(result_type(0.0, a.dtype) for a in arrays))`...)

Reply:
@seberg 

* `np.result_type(0.0, np.int16, np.float32)` gives `float32` for promotion state `legacy` and `float64` for promotion state `weak`. But that change is _unrelated_ to this or any of the other PRs (as the PRs do not change the `result_type` itself)

* For `common_type` the output of `common_type(i16, f32)` is `float64` on main (for both weak and legacy promotion rules). For this PR the output of `common_type(i16, f32)` is `float32` (a behavior change). For #24668 the output is `float64` (so no behaviour change

Summary:

* #24467: Pure optimization, no behaviour changes at all
* #24656: Fastest option, uses `result_type` internally, behaviour change for `i16, f32`, does broaden the signature
* #24668: Uses `result_type` internally, which simplifies code. Only behaviour change observed is for datetime64

Reply:
> np.result_type(0.0, np.int16, np.float32) gives float32 for promotion state legacy and float64 for promotion state weak.

what the heck!  That looks wrong, that result shouldn't change with weak promotion!  Some code path is mistriggring and thinking this cannot have any weak promotion at all or so.

Reply:
@seberg Not sure whether this helps, but the output depends on how the `i16` and `f32` are passed (as numpy scalar, numpy array or dtype):
```
import numpy as np
print(np, np.__version__)

i16scalar = np.int16(2)
f32scalar = np.float32(3.)

i16array = np.int16([2, 2])
f32array = np.float32([3., 3.])

for s in ['weak', 'legacy']:
    print(s)
    np._set_promotion_state(s)
    print('np.result_type(0.0, i16scalar, f32scalar)', np.result_type(0.0, i16scalar, f32scalar))
    print('np.result_type(0.0, i16array, f32array)', np.result_type(0.0, i16array, f32array))
    print('np.result_type(0.0, i16.dtype, f32.dtype)', np.result_type(0.0, i16scalar.dtype, f32scalar.dtype))
    print('np.result_type(0.0, f32)', np.result_type(0.0, f32scalar))
```
Output
```
import numpy as np
print(np, np.__version__)

i16scalar = np.int16(2)
f32scalar = np.float32(3.)

i16array = np.int16([2, 2])
f32array = np.float32([3., 3.])

for s in ['weak', 'legacy']:
    print(s)
    np._set_promotion_state(s)
    print('np.result_type(0.0, i16scalar, f32scalar)', np.result_type(0.0, i16scalar, f32scalar))
    print('np.result_type(0.0, i16array, f32array)', np.result_type(0.0, i16array, f32array))
    print('np.result_type(0.0, i16.dtype, f32.dtype)', np.result_type(0.0, i16scalar.dtype, f32scalar.dtype))
    print('np.result_type(0.0, f32)', np.result_type(0.0, f32scalar))
```

Reply:
@seberg I am going to close this PR in favor of #24668 (which includes your two suggestions). The performance of this PR is a better, but maybe not worth the downsides. The only usage of `np.common_type` is in `np.polynomial.polyutils` and there we could just inline `np.result_type` if needed.

I wrote earlier `np.result_type(0.0, np.int16, np.float32) gives float32 for promotion state legacy and float64 for promotion state weak`, but actually it is the other way around. For posterity: on current main these are the promotion rules
```
import numpy as np
print(np, np.__version__)

i16scalar = np.int16(2)
f32scalar = np.float32(3.)

i16array = np.int16([2, 2])
f32array = np.float32([3., 3.])

for s in ['weak', 'legacy']:
    print(f'\n\n### promotion {s}')
    np._set_promotion_state(s)
    print('np.result_type(i16scalar, f32scalar)', np.result_type(i16scalar, f32scalar))
    print('np.result_type(i16array, f32array)', np.result_type( i16array, f32array))
    print('np.result_type(i16.dtype, f32.dtype)', np.result_type( i16scalar.dtype, f32scalar.dtype))
    print('np.result_type(f32)', np.result_type(f32scalar))
    
    print('np.result_type(0.0, i16scalar, f32scalar)', np.result_type(0.0, i16scalar, f32scalar))
    print('np.result_type(0.0, i16array, f32array)', np.result_type(0.0, i16array, f32array))
    print('np.result_type(0.0, i16.dtype, f32.dtype)', np.result_type(0.0, i16scalar.dtype, f32scalar.dtype))
    print('np.result_type(0.0, f32)', np.result_type(0.0, f32scalar))
    
    print('np.result_type(0.0, np.int32, np.float32)', np.result_type(0.0, np.int32, np.float32))

    print()
    print('np.result_type(0., i16scalar, f32scalar)', np.result_type(0.,  i16scalar, f32scalar))
    print('np.result_type(0., i16scalar.dtype, f32scalar.dtype)', np.result_type( 0., i16scalar.dtype, f32scalar.dtype))

    print('np.result_type(0.,i16array, f32array)', np.result_type(0.,  i16array, f32array))
    print('np.result_type(0.,i16array.dtype, f32array.dtype)', np.result_type( 0., i16array.dtype, f32array.dtype))
```
Results in:
```

### promotion weak
np.result_type(i16scalar, f32scalar) float32
np.result_type(i16array, f32array) float32
np.result_type(i16.dtype, f32.dtype) float32
np.result_type(f32) float32
np.result_type(0.0, i16scalar, f32scalar) float32
np.result_type(0.0, i16array, f32array) float32
np.result_type(0.0, i16.dtype, f32.dtype) float32
np.result_type(0.0, f32) float32
np.result_type(0.0, np.int32, np.float32) float64

np.result_type(0., i16scalar, f32scalar) float32
np.result_type(0., i16scalar.dtype, f32scalar.dtype) float32
np.result_type(0.,i16array, f32array) float32
np.result_type(0.,i16array.dtype, f32array.dtype) float32


### promotion legacy
np.result_type(i16scalar, f32scalar) float32
np.result_type(i16array, f32array) float32
np.result_type(i16.dtype, f32.dtype) float32
np.result_type(f32) float32
np.result_type(0.0, i16scalar, f32scalar) float64
np.result_type(0.0, i16array, f32array) float32
np.result_type(0.0, i16.dtype, f32.dtype) float32
np.result_type(0.0, f32) float64
np.result_type(0.0, np.int32, np.float32) float64

np.result_type(0., i16scalar, f32scalar) float64
np.result_type(0., i16scalar.dtype, f32scalar.dtype) float32
np.result_type(0.,i16array, f32array) float32
np.result_type(0.,i16array.dtype, f32array.dtype) float32
```